spec/predicates/runtime-trace.md

```diff
@@ -0,0 +1,126 @@
+# Predicate type: Runtime Trace
+
+Type URI: https://in-toto.io/runtime-trace/v0.1.0
+
+Version: 0.1.0
+
+## Purpose
+
+This predicate is used to describe system events that were part of the artifact build process. Artifact build processes today are complex and opaque with little to no insights into build actions. However, there are emerging introspection techniques that now allow us to monitor and capture system event logs during the build.
+
+## Prerequisites
+
+Understanding of monitoring tools and the larger in-toto attestation framework.
+
+## Model
+
+The structure of this predicate closely mirrors the structure of the SLSA Provenance attestation as they are complementary. `monitor` identifies the specific instance of a tool that’s observing the build, while `build` contains information that when put together uniquely identifies the exact instance of the job being observed. `invocation` details the configurations of `monitor` and the `monitorLog` contains the actual runtime trace information.
+
+This attestation can be used in conjunction with provenance attestations to leverage observability to attest to various SLSA requirements. The runtime trace can prove the build was invoked via a script, that the build was executed in a hermetic environment with no network access, and so on.
+
+## Schema
+
+The schema of the attestation type is the core part of the document. It defines the fields included in an instantiation of the attestation type, as well as rules to parse them.
+
+```
+{
+  "_type": "https://in-toto.io/Statement/v0.1",
+  "predicateType": "https://in-toto.io/runtime-trace/v0.1.0",
+  "subject": [{ ... }],
+  "predicate": {
+    "monitor": {
+      "id": "<URI>"
+    },
+    “build”: {
+        “BuilderId”: “<URI>”,
+        “type”: “<URI>”,
+        “event”: “<STRING>”,
+    },
+    "invocation": {
+     "configSource": {
+        "uri": "<URI>",
+        "digest": { /* DigestSet */ },
+        "entryPoint": "<STRING>"
+      },
+      "tracePolicy": {}
+    },
+    "monitorLog": {
+      "process": [
+         { /* object */ }
```

Comment:
Could we define "object" better here?  Normalizing output from various trace tools will be helpful in discoverability.

I have some rough ideas of that could look like in this branch.  https://github.com/testifysec/go-witness/blob/tetragon/attestation/commandrun/commandrun.go

Reply:
Hey Cole,

yes we have defied it in the demo example as the following. Would be great to get feedback :
``` {
          "eventType": "kprobe",
          "functionName": “__x64_sys_write/fd_install”,
          “processBinary” : “/bin/cat”
          “exitCode”: int
          "arguments": [
               "-wait_file-post_file ",
               "/tekton/run/2/out",
               "/tekton/run/3/out",
               "-termination_path",
               "/tekton/termination",
               "-step_metadata_dir",
               "/tekton/run/3/status",
               "-entrypoint",
               "/tekton/scripts/script-3-g5chp",
               "--",
          ],
          "privileged": [
                 “CAP_SYS_ADMIN”,
                 ],
          },
          {
          "eventType": "kprobe",
          "functionName": “tcp_connect”,
          “processBinary” : “”
          "arguments": [
               "tcp",
               "10.244.0.6:34965",
               "104.198.14.52:8",
          ],
          "privileged": [],
    },
```
we do want to add hash of files and such as that might be useful. Looking at your branch we are very much on the same page and would love to work together on it. I think there is information that you include that would be valuable to add into the predicate.

Reply:
We will probably adopt whatever schema is defined here.  The only issue we have is that the data needs to be normalized.  For our use case, we would like the user to be able to define the same policy no matter what trace tool was used.  

It looks like the schema here is from Tetragon.  I think this needs to be mutated into a comment format, and we should add another field(s) to define the trace provider.

Reply:
100% agree. We do want to make this generic that other tools can be used. Adding in fields that are missing. As for trace provider we do have:
```
  "predicate": {
    "monitor": {
      "id": "<URI>"
    },
```
which would define which tool is being used.